File: scripts/import_interactions.py
import firebase_admin
from firebase_admin import credentials, firestore
import csv
import os
import time
from pathlib import Path

# Initialize Firebase Admin
cred = credentials.Certificate('smartpilladvisor-firebase-adminsdk-fbsvc-5ddf1b35d7.json')
firebase_admin.initialize_app(cred)

db = firestore.client()

# Path to CSV files
csv_dir = Path(__file__).parent.parent / 'ddinterpy'

csv_files = [
    'ddinter_downloads_code_A.csv',
    'ddinter_downloads_code_B.csv',
    'ddinter_downloads_code_D.csv',
    'ddinter_downloads_code_H.csv',
    'ddinter_downloads_code_L.csv',
    'ddinter_downloads_code_P.csv',
    'ddinter_downloads_code_R.csv',
    'ddinter_downloads_code_V.csv'
]

def import_csv_to_firestore():
    total_count = 0
    batch = db.batch()
    batch_count = 0
    
    print("Starting import...")
    
    for csv_file in csv_files:
        file_path = csv_dir / csv_file
        
        if not file_path.exists():
            print(f"❌ File not found: {file_path}")
            continue
            
        print(f"\n📄 Processing {csv_file}...")
        file_count = 0
        
        with open(file_path, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            
            for row in reader:
                doc_ref = db.collection('interactions').document()
                batch.set(doc_ref, {
                    'Drug_A': row['Drug_A'],
                    'Drug_B': row['Drug_B'],
                    'Level': row['Level'],
                    'DDInterID_A': row.get('DDInterID_A') or None,
                    'DDInterID_B': row.get('DDInterID_B') or None
                })
                
                batch_count += 1
                file_count += 1
                total_count += 1
                
                # Firestore batch limit is 500, use smaller batches to avoid quota
                if batch_count >= 100:
                    try:
                        batch.commit()
                        print(f"  ✓ Committed batch at {total_count} records")
                        batch = db.batch()
                        batch_count = 0
                        time.sleep(1)  # Wait 1 second between batches to avoid quota
                    except Exception as e:
                        if "Quota exceeded" in str(e) or "RESOURCE_EXHAUSTED" in str(e):
                            print(f"  ⚠️ Quota limit hit. Waiting 10 seconds...")
                            time.sleep(10)
                            batch.commit()
                            batch = db.batch()
                            batch_count = 0
                        else:
                            raise
        
        print(f"  ✅ {csv_file}: {file_count} records")
    
    # Commit remaining batch
    if batch_count > 0:
        batch.commit()
        print(f"  ✓ Committed final batch")
    
    print(f"\n🎉 Import complete!")
    print(f"📊 Total records imported: {total_count}")
    print(f"\n✅ Check Firestore Console: https://console.firebase.google.com/project/smartpilladvisor/firestore")

if __name__ == '__main__':
    try:
        import_csv_to_firestore()
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()

File: scripts/importInteractions.js
/* eslint-disable @typescript-eslint/no-require-imports */
// Script to import drug interactions from CSV files to Firestore
const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

// Initialize Firebase Admin
const serviceAccount = require('../firebase-service-account.json'); // You'll need to download this

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  projectId: 'smartpilladvisor'
});

const db = admin.firestore();

// Path to CSV files
const CSV_DIR = 'C:\\Users\\Yugendra\\Downloads\\Drug-Interaction-Checking-Website-main\\Drug-Interaction-Checking-Website-main\\ddinterpy';

const csvFiles = [
  'ddinter_downloads_code_A.csv',
  'ddinter_downloads_code_B.csv',
  'ddinter_downloads_code_D.csv',
  'ddinter_downloads_code_H.csv',
  'ddinter_downloads_code_L.csv',
  'ddinter_downloads_code_P.csv',
  'ddinter_downloads_code_R.csv',
  'ddinter_downloads_code_V.csv'
];

async function importInteractions() {
  console.log('Starting drug interaction import...');
  let totalCount = 0;
  let batch = db.batch();
  let batchCount = 0;

  for (const csvFile of csvFiles) {
    const filePath = path.join(CSV_DIR, csvFile);
    console.log(`Processing ${csvFile}...`);

    await new Promise((resolve, reject) => {
      fs.createReadStream(filePath)
        .pipe(csv())
        .on('data', (row) => {
          // Create document in Firestore
          const docRef = db.collection('interactions').doc();
          batch.set(docRef, {
            Drug_A: row.Drug_A,
            Drug_B: row.Drug_B,
            Level: row.Level,
            DDInterID_A: row.DDInterID_A || null,
            DDInterID_B: row.DDInterID_B || null
          });

          batchCount++;
          totalCount++;

          // Firestore batch limit is 500
          if (batchCount >= 450) {
            batch.commit();
            batch = db.batch();
            batchCount = 0;
            console.log(`  Committed batch. Total so far: ${totalCount}`);
          }
        })
        .on('end', () => {
          console.log(`  Finished ${csvFile}`);
          resolve();
        })
        .on('error', reject);
    });
  }

  // Commit remaining batch
  if (batchCount > 0) {
    await batch.commit();
    console.log(`  Committed final batch`);
  }

  console.log(`\n✅ Import complete! Total interactions imported: ${totalCount}`);
  process.exit(0);
}

importInteractions().catch((error) => {
  console.error('Error importing interactions:', error);
  process.exit(1);
});
